### PR TITLE
[Proof] refactor: ring client for on-chain use

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

1. Factor out a `RingClient` from the `RingCache`.
2. Add `AppKeeperQueryClient` and `AccountKeeperQueryClient` wrappers to implement `ApplicationQueryClient` and `AccountQueryClient`, respectively.
3. Refactor the proof keeper to instantiate and use a `RingClient` to implement `#GetRingForAddress()`.

### AI Summary

reviewpad:summary

## Issue

- #406 
- #401 

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/proxy/errors.go [hunk 1]
```diff
@@ -5,14 +5,12 @@ import sdkerrors "cosmossdk.io/errors"
 var (
 	codespace                                        = "relayer_proxy"
 	ErrRelayerProxyUnsupportedRPCType                = sdkerrors.Register(codespace, 1, "unsupported relayer proxy rpc type")
-	ErrRelayerProxyInvalidRelayRequestSignature      = sdkerrors.Register(codespace, 2, "invalid relay request signature")
 	ErrRelayerProxyInvalidSession                    = sdkerrors.Register(codespace, 3, "invalid session in relayer request")
```
Comment (bryanchriswhite) [inline pkg/relayer/proxy/errors.go:8]: Looks like we need to re-sequence the error codes.

File: pkg/crypto/interface.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ import (
 	"context"
 
 	"github.com/noot/ring-go"
+	"github.com/pokt-network/poktroll/x/service/types"
```
Comment (bryanchriswhite) [inline pkg/crypto/interface.go:8]:
```suggestion

	"github.com/pokt-network/poktroll/x/service/types"
```

File: pkg/crypto/rings/client.go [hunk 3]
```diff
@@ -69,6 +72,84 @@ func (rc *ringClient) GetRingForAddress(
...
+			"request payload: %s", relayRequest.Payload,
+		)
+	}
+
+	signature := relayRequest.Meta.Signature
+	if signature == nil {
+		return sdkerrors.Wrapf(
+			ErrRingClientInvalidRelayRequest,
+			"missing signature from relay request: %v", relayRequest,
+		)
```
Comment (bryanchriswhite) [inline pkg/crypto/rings/client.go:101]: Let's rewrite these error constructors to be less verbose:

```suggestion
		return ErrRingClientInvalidRelayRequest.Wrapf(
			"missing signature from relay request: %v", relayRequest,
		)
```

File: pkg/relayer/proxy/relay_verifier.go [hunk 4]
```diff
@@ -18,74 +14,14 @@ func (rp *relayerProxy) VerifyRelayRequest(
...
-		)
+	if err := rp.ringCache.VerifyRelayRequestSignature(ctx, relayRequest); err != nil {
+		return err
 	}
 
-	// get the ring for the application address of the relay request
+	// Application address is used to verify the relayRequest signature, it is
+	// guaranteed to be present in the relayRequest since the signature has already
+	// been verified.
 	appAddress := relayRequest.Meta.SessionHeader.ApplicationAddress
```
Comment (bryanchriswhite) [inline pkg/relayer/proxy/relay_verifier.go:24]: nit: can we use the getter methods here?

File: x/proof/keeper/msg_server_submit_proof.go [hunk 5]
```diff
@@ -108,8 +106,7 @@ func (k msgServer) SubmitProof(ctx context.Context, msg *types.MsgSubmitProof) (
 	}
 
 	// Verify the relay request's signature.
-	appAddress := msg.GetSessionHeader().GetApplicationAddress()
-	if err := k.verifyRelayRequestSignature(ctx, relay.GetReq(), appAddress); err != nil {
+	if err := k.ringClient.VerifyRelayRequestSignature(ctx, relay.GetReq()); err != nil {
```
Comment (bryanchriswhite) [inline x/proof/keeper/msg_server_submit_proof.go:109]: :ok_hand:

**Review by red-0ne — APPROVED** [review]

**Pull request merged**